reconcile/oum/standalone.py

```diff
@@ -0,0 +1,197 @@
+        actionable to them.
+        """
+        logging.warning(
+            "Failed to reconcile cluster user group configuration on "
+            f"cluster {spec.cluster.ocm_cluster.name} (id={spec.cluster.ocm_cluster.id}) "
+            f"in the OCM organization {spec.cluster.organization_id}",
+            exc_info=error,
+        )
+
+
+def user_mgmt_label_key(config_atom: str) -> str:
+    """
+    Generates label keys for the rover cluter group authz capability, compliant with the naming
+    scheme defined in https://service.pages.redhat.com/dev-guidelines/docs/sre-capabilities/framework/ocm-labels/
+    """
+    return sre_capability_label_key("user-mgmt", config_atom)
+
+
+def discover_clusters(
+    ocm_api: OCMBaseClient,
+    org_ids: Optional[set[str]] = None,
+) -> dict[str, list[ClusterDetails]]:
+    """
+    Discover all clusters with user management enabled on their subscription
+    or their organization. Return the discovered clusters grouped by OCM organization ID.
+
+    If `org_ids` is provided, only clusters from the specified organizations will be returned.
+    """
+    clusters = discover_clusters_by_labels(
+        ocm_api=ocm_api,
+        label_filter=Filter().like("key", user_mgmt_label_key("%")),
+    )
+
+    # group by org ID
+    # optionally also filter on org IDs
+    clusters_by_org: dict[str, list[ClusterDetails]] = defaultdict(list)
+    for c in clusters:
+        passed_ocm_filters = org_ids is None or c.organization_id in org_ids
+        if passed_ocm_filters:
+            clusters_by_org[c.organization_id].append(c)
+
+    return clusters_by_org
+
+
+def build_user_management_configurations(
+    org_id: str,
+    clusters: list[ClusterDetails],
+    providers: set[str],
+) -> OrganizationUserManagementConfiguration:
+    """
+    Extracts the user management configuration from the cluster labels.
+    """
+    org_config = OrganizationUserManagementConfiguration(org_id=org_id)
+    for c in clusters:
+        cluster_config = ClusterUserManagementConfiguration(cluster=c)
+        org_config.cluster_configs.append(cluster_config)
+        for p in providers:
+            provider_org_labels = build_container_for_prefix(
+                c.organization_labels, user_mgmt_label_key(f"{p}."), True
+            )
+            provider_subs_labels = build_container_for_prefix(
+                c.subscription_labels, user_mgmt_label_key(f"{p}."), True
+            )
+            if provider_org_labels or provider_subs_labels:
+                try:
+                    role_groups = build_cluster_config_from_labels(
+                        provider=p,
+                        org_labels=provider_org_labels,
+                        subscription_labels=provider_subs_labels,
+                    )
+                    for role_id, group_refs in role_groups.items():
+                        cluster_config.roles[role_id].extend(group_refs)
+
+                except Exception as e:
```

Comment:
What exception are we expecting here?

Reply:
pydantic validation errors. the `build_cluster_config_from_labels` processes OCM labels, which are basically free form labels. a pydantic dataclass takes care of validating some formal constraints and fails if they are not met.

i can for sure be more specific here by using the concrete pydantic error class. on the other hand i don't want any potential error from within this block to leak out ever.

Reply:
Right, that pydantic validation, I thought there were some external calls, but didn't find any, good to catch them!